feat(read): perform reading from storage in parallel for faster reads

lionel1704 opened the pull request:
Note that this PR also contains some changes to allow parallel writes, but this is not supported by sn_client at the moment and hence it is disabled.

maqi commented on src/self_encryptor.rs:
```diff
@@ -266,6 +267,7 @@ where
             }
         }
 
+        // let mut network_storage_futures = vec![];
```
forget to remove?

maqi commented on src/self_encryptor.rs:
```diff
@@ -281,16 +283,21 @@ where
+                // network_storage_futures.push(async move { storage.put(name.to_vec(), content).await });
             }
         }
+        // let results = join_all(network_storage_futures.into_iter()).await;
```
removed?

maqi commented on src/self_encryptor.rs:
```diff
@@ -343,11 +352,15 @@ where
+    let decrypted_data = join_all(decryption_futures).await;
+    let mut pos_iter = positions.into_iter();
+    for chunk in decrypted_data {
+        decrypted_chunks.push((chunk?, pos_iter.next().unwrap_or(0)))
```
`any` of failed decryption will cause `all left over` decrypted chunk to be dropped ?

maqi commented on src/self_encryptor.rs:
```diff
@@ -508,16 +521,23 @@ where
+    let chunks = join_all(decryption_futures.into_iter()).await;
+    let mut pos_iter = positions.into_iter();
+    for chunk in chunks {
+        decrypted_chunks.push((chunk.unwrap(), pos_iter.next().unwrap_or(0)))
```
use `?` or `unwrap` ?

dirvine replied:
Not unwrap of course ;-)

maqi commented on src/sequential/large_encryptor.rs:
```diff
@@ -143,7 +159,7 @@ where
         data = self.fill_chunk_buffer(data, 1).await?;
 
-        let mut all_things = Vec::new();
+        // let mut storage_futures = Vec::new();
```
remove?

maqi commented on src/sequential/large_encryptor.rs:
```diff
@@ -156,9 +172,14 @@ where
+                // storage_futures.push(self.encrypt_chunk(&data_to_encrypt, index).await?);
             }
         }
+        // let results = futures::future::join_all(storage_futures.into_iter()).await;
```
remove?

maqi commented on src/sequential/medium_encryptor.rs:
```diff
@@ -18,31 +20,38 @@ pub const MAX: usize = 3 * MAX_CHUNK_SIZE;
+        let results = join_all(get_futures.into_iter()).await;
+        for chunk in results {
+            let (chunk, pad_key_iv) = chunk?;
             let decrypted_chunk = utils::decrypt_chunk(&chunk, pad_key_iv)?;
```
can this `decrypt` operation to be moved into the `futures` as well? i.e. the result is `decrypted_chunk` directly?
i.e. the async function does `get` and `decrypt` in one go

maqi commented on src/sequential/medium_encryptor.rs:
```diff
@@ -78,7 +87,7 @@ where
         }
 
-        let mut all_chunks;
+        // let mut chunk_storage_futures;
```
remove?

maqi commented on src/sequential/medium_encryptor.rs:
```diff
@@ -113,9 +122,15 @@ where
                 details.hash = hash.to_vec();
-                all_chunks.push(self.storage.put(hash.to_vec(), encrypted_contents).await?);
+                self.storage.put(hash.to_vec(), encrypted_contents).await?
+                // let mut storage = self.storage.clone();
```
remove?

maqi commented on src/test_helpers.rs:
```diff
@@ -40,27 +41,39 @@ impl<'a> Debug for Blob<'a> {
+        Ok(self
+            .entries
+            .read()
+            .map_err(|_| SelfEncryptionError::Poison)?
```
better at least log the error?

lionel1704 replied:
This error happens for test only and there is only one possible error. So I think we should be okay here.

lionel1704 commented:
@maqi I've push a comment addressing your comments. The code that is commented out is left there on purpose. As I've mentioned in the PR description, it meant for the parallel writes which is not yet supported by sn_client. They will be un-commented once that is ready.
